rfcs/0062-content-addressed-paths.md

```diff
@@ -108,118 +156,139 @@ create table if not exists Realisation (
+For a derivation `drv` whose input derivations have all been realised, we define its **associated resolved derivation** `resolved(drv)` as `drv` in which
+we replace every input derivation `inDrv` of `drv` by `Realisation(inDrv).path`.
 
 `resolved` is (intentionally) not injective: If `drv` and `drv'` only differ because one depends on `dep` and the other on `dep'`, but `dep` and `dep'` are content-addressed and have the same output hash, then `resolved(drv)` and `resolved(drv')` will be equal.
 
-#### Build process
-
-When asked to build a derivation `drv`, we instead:
-
-1. Compute `resolved(drv)`
-2. Substitute and build `resolved(drv)` like a normal derivation.
-   Possibly this is a no-op because it may be that `resolved(drv)` has already been built.
-3. Add a new mapping `Realisation(drv!${output}) == ${output}(resolved(drv))` for each output `output` of `drv` (signing the mapping if needs be)
-
-### Building a CA derivation
-
-A **CA derivation** is a derivation with the `__contentAddressed` argument set
-to `true` and the `outputHashAlgo` set to a value that is a valid hash name
-recognized by Nix (see the description for `outputHashAlgo` at
-<https://nixos.org/nix/manual/#sec-advanced-attributes> for the current allowed
-values).
-
-The process for building a content-adressed derivation `drv` is the following:
-
-- We build it like a normal derivation (see above).
-  For each output `$outputId` of the derivation, this gives us a (temporary) output path `$out`.
-  - We compute a cryptographic hash `$chash` of `$out`[^modulo-hashing]
-  - We move `$out` to `/nix/store/$chash-$name`
-  - We store the mapping `Realisation($outputId) == "/nix/store/$chash-$name"`
-
-[^modulo-hashing]:
-
-  We can possibly normalize all the self-references before
-  computing the hash and rewrite them when moving the path to handle paths with
-  self-references, but this isn't strictly required for a first iteration
-
-### Example
-
-In this example, we have the following Nix expression:
-
-```nix
-rec {
-  contentAddressed = mkDerivation {
-    name = "contentAddressed";
-    __contentAddressed = true;
-    … # Some extra arguments
-  };
-  dependent = mkDerivation {
-    name = "dependent";
-    buildInputs = [ contentAddressed ];
-    … # Some extra arguments
-  };
-  transitivelyDependent = mkDerivation {
-    name = "transitivelyDependent";
-    buildInputs = [ dependent ];
-    … # Some extra arguments
-  };
-}
+### content-addressed build process
+
+We now need to update the build process as:
+
+```python
+def build_derivation(derivation : Derivation, outputsToBuild: [str]) -> Map[DrvOutput, Realisation]:
+    inputRealisations : Map[DrvOutput, Realisation] = {}
+    # Build all the inputs, and store the newly built realisations
+    for (inputDrv, requiredOutputs) in derivation.inputDrvs:
+        inputRealisations += build_derivation(inputDrv, requiredOutputs)
```

Comment:
this line is a bit confusing... where does the key for the new entry in `inputRealisations`, which is a associative array afaik, come from?

Reply:
`build_derivation` also returns a map. Would something like

```python
inputRealisations.extend(build_derivation(inputDrv, requiredOutputs))
```

be clearer?

Reply:
yes, thanks.